### DISCUSSION TITLE
Update statement on MSSql does not support like operator

### DISCUSSION THREAD

**Discussion by paoloalba** [sqlalchemy/sqlalchemy#7692] [Usage Questions] [answered]
### Describe the bug

I try to update an MSSql database through an update statement. When I use a "like" operator among my where clauses, I get an exception.

### To Reproduce

```python
tbl_obj = getattr(self.base.classes, table_name)
where_list = []
where_list.append()

filter_column = getattr(tbl_obj, column_name)

where_list.append(filter_column.contains(match_elem))

stmt = update(tbl_obj).where(and_(*where_list)).values(update_dict)
result = self.session.execute(stmt)
self.session.commit()
```

### Error

---------------------------------------------------------------------------
UnevaluatableError                        Traceback (most recent call last)
File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py:1938, in BulkUDCompileState._do_pre_synchronize_evaluate(cls, session, statement, params, execution_options, bind_arguments, update_options)
   1937 if crit:
-> 1938     eval_condition = evaluator_compiler.process(*crit)
   1939 else:

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/evaluator.py:85, in EvaluatorCompiler.process(self, *clauses)
     82     raise UnevaluatableError(
     83         "Cannot evaluate %s" % type(clause).__name__
     84     )
---> 85 return meth(clause)

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/evaluator.py:136, in EvaluatorCompiler.visit_clauselist(self, clause)
    135 def visit_clauselist(self, clause):
--> 136     evaluators = list(map(self.process, clause.clauses))
    137     if clause.operator is operators.or_:

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/evaluator.py:85, in EvaluatorCompiler.process(self, *clauses)
     82     raise UnevaluatableError(
     83         "Cannot evaluate %s" % type(clause).__name__
     84     )
---> 85 return meth(clause)

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/evaluator.py:214, in EvaluatorCompiler.visit_binary(self, clause)
    213 else:
--> 214     raise UnevaluatableError(
    215         "Cannot evaluate %s with operator %s"
    216         % (type(clause).__name__, clause.operator)
    217     )
    218 return evaluate

UnevaluatableError: Cannot evaluate BinaryExpression with operator <function like_op at 0x7fb56ace2b80>

The above exception was the direct cause of the following exception:

InvalidRequestError                       Traceback (most recent call last)

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/session.py:1640, in Session.execute(self, statement, params, execution_options, bind_arguments, _parent_execute_state, _add_event, **kw)
   1634 execution_options = util.coerce_to_immutabledict(execution_options)
   1636 if compile_state_cls is not None:
   1637     (
   1638         statement,
   1639         execution_options,
-> 1640     ) = compile_state_cls.orm_pre_session_exec(
   1641         self,
   1642         statement,
   1643         params,
   1644         execution_options,
   1645         bind_arguments,
   1646         _parent_execute_state is not None,
   1647     )
   1648 else:
   1649     bind_arguments.setdefault("clause", statement)

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py:1817, in BulkUDCompileState.orm_pre_session_exec(cls, session, statement, params, execution_options, bind_arguments, is_reentrant_invoke)
   1809 # this stage of the execution is called before the do_orm_execute event
   1810 # hook.  meaning for an extension like horizontal sharding, this step
   1811 # happens before the extension splits out into multiple backends and
   1812 # runs only once.  if we do pre_sync_fetch, we execute a SELECT
   1813 # statement, which the horizontal sharding extension splits amongst the
   1814 # shards and combines the results together.
   1816 if update_options._synchronize_session == "evaluate":
-> 1817     update_options = cls._do_pre_synchronize_evaluate(
   1818         session,
   1819         statement,
   1820         params,
   1821         execution_options,
   1822         bind_arguments,
   1823         update_options,
   1824     )
   1825 elif update_options._synchronize_session == "fetch":
   1826     update_options = cls._do_pre_synchronize_fetch(
   1827         session,
   1828         statement,
   (...)
   1832         update_options,
   1833     )

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py:1945, in BulkUDCompileState._do_pre_synchronize_evaluate(cls, session, statement, params, execution_options, bind_arguments, update_options)
   1942             return True
   1944 except evaluator.UnevaluatableError as err:
-> 1945     util.raise_(
   1946         sa_exc.InvalidRequestError(
   1947             'Could not evaluate current criteria in Python: "%s". '
   1948             "Specify 'fetch' or False for the "
   1949             "synchronize_session execution option." % err
   1950         ),
   1951         from_=err,
   1952     )
   1954 if statement.__visit_name__ == "lambda_element":
   1955     # ._resolved is called on every LambdaElement in order to
   1956     # generate the cache key, so this access does not add
   1957     # additional expense
   1958     effective_statement = statement._resolved

File /opt/conda/lib/python3.9/site-packages/sqlalchemy/util/compat.py:207, in raise_(***failed resolving arguments***)
    204     exception.__cause__ = replace_context
    206 try:
--> 207     raise exception
    208 finally:
    209     # credit to
    210     # https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/
    211     # as the __traceback__ object creates a cycle
    212     del exception, replace_context, from_, with_traceback

InvalidRequestError: Could not evaluate current criteria in Python: "Cannot evaluate BinaryExpression with operator <function like_op at 0x7fb56ace2b80>". Specify 'fetch' or False for the synchronize_session execution option.

### Versions

- OS: Ubuntu 20.04.3 LTS
- Python:3.9.7
- SQLAlchemy:1.4.31
- Database: MSSql - SQL Server 12.0.2000.8
- DBAPI: pyodbc 4.0.32

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's expected. Like the exception suggests you should specify a synchronize_session mode. See https://docs.sqlalchemy.org/en/14/orm/session_basics.html#orm-expression-update-delete for more details
